File: Recursion/ReverseString.py
from typing import List

class Solution:
    def __init__(self , s:List[str]):
        self.reverseString(s)

    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        if(len(s)==0):
            return
        temp = s[0]
        s.pop(0)
        self.reverseString(s)
        s.append(temp)

if __name__=='__main__':
    _list = ['h' , 'e' , 'l' , 'l' , 'o']
    obj = Solution(_list)
    print(_list)

        

File: Recursion/ReverseString.cpp
// https://leetcode.com/explore/featured/card/recursion-i/250/principle-of-recursion/1440/
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    void reverseString(vector<char>& s) {
        if (s.size()==0) return;
        char temp = s[0];
        s.erase(s.begin());
        reverseString(s);
        s.push_back(temp);
    }
};

int main(){
    vector<char> s ={'H' , 'a' , 'n' , 'n' , 'a' , 'h'};
    Solution obj;
    obj.reverseString(s);
    for(auto i:s) cout << i << ' ';
    cout << '\n';
    return 0;
}